### PR TITLE
Update PoW Difficulty

### LINKED ISSUE #18: Implement Market Maker PoW Difficulty
Implement market maker PoW difficulty so that we can upgrade the existing PoW contract in band.

### PR DESCRIPTION
**Pull request by mvandeberg**
Closes #18

### CODE REVIEW

File: contracts/pow/pow.cpp [hunk 1]
```diff
@@ -23,70 +29,185 @@ struct pow_signature_data
...
+   y >>= 32;
+   y += 0x100000000;
+   if( y < 0 )
+      y = 0;
+   return uint64_t( y );
+   /*
+   y = (y > 0) ? 0 : -y;
+   return uint64_t( y );
+   */
+}
```
Comment (theoreticalbts) [inline contracts/pow/pow.cpp:92]: Let's remove these commented-out lines.

File: contracts/pow/pow.cpp [hunk 2]
```diff
@@ -23,70 +29,185 @@ struct pow_signature_data
...
+   // x*y = k
+   // -> (x+dx)*(y+dy) = x*y
+   // -> x*y + dx*y + dy*x + dx*dy = x*y
+   // -> dx*(y+dy) = -dy*x
+   // -> dx = -dy*x/(y+dy)
+   //
+   // dy = -TARGET_BLOCK_INTERVAL_MS
+
+   // TODO:  Overflow analysis
+
```
Comment (theoreticalbts) [inline contracts/pow/pow.cpp:111]: This TODO comment needs a corresponding issue.

File: contracts/pow/pow.cpp [hunk 3]
```diff
@@ -23,70 +29,185 @@ struct pow_signature_data
...
+   diff_meta.required_hc = uint64_t( ((1000 * TARGET_BLOCK_INTERVAL_MS) * diff_meta.hc_reserve) / (diff_meta.us_reserve - (1000 * TARGET_BLOCK_INTERVAL_MS)) );
+   ++diff_meta.required_hc;
+   diff_meta.difficulty_target = std::numeric_limits< uint256_t >::max() / diff_meta.required_hc;
+}
+
+difficulty_metadata initialize_difficulty( const difficulty_metadata_v1& pow1_meta )
+{
+   uint64_t initial_hashes_per_block = ( std::numeric_limits< uint256_t >::max() / pow1_meta.difficulty_target ).convert_to< uint64_t >();
+   uint64_t initial_hashes_per_second = initial_hashes_per_block / ( TARGET_BLOCK_INTERVAL_MS / 1000 );
+   uint64_t effective_us = uint64_t( 1000 ) * TARGET_BLOCK_INTERVAL_MS;
```
Comment (theoreticalbts) [inline contracts/pow/pow.cpp:121]: This computation should be `initial_hashes_per_second = (1000 * initial_hashes_per_block) / TARGET_BLOCK_INTERVAL_MS;` so it Does the Right Thing even if `TARGET_BLOCK_INTERVAL_MS` is not evenly divisible by 1000.

However, if we do this, we also need an overflow check to be sure the multiplication in the numerator doesn't overflow.  Which can go in the same overflow analysis TODO issue as other overflow checks in this smart contract.

File: contracts/pow/pow.cpp [hunk 4]
```diff
@@ -23,70 +29,185 @@ struct pow_signature_data
...
+   (difficulty_target)
+   (last_block_time)
+   (required_hc)
+   (hc_reserve)
+   (us_reserve)
+   (target_block_interval)
+)
+
+using difficulty_metadata = difficulty_metadata_v2;
+
```
Comment (theoreticalbts) [inline contracts/pow/pow.cpp:67]: This `using difficulty_metadata` should be eliminated, and replace `difficulty_metadata` with `difficulty_metadata_v2` in the following code.  The `using` makes the code a little harder to follow than simply being explicit everywhere.

File: contracts/pow/pow.cpp [hunk 5]
```diff
@@ -23,70 +29,185 @@ struct pow_signature_data
...
+ */
+uint64_t get_effective_us(uint64_t dt_ms)
+{
+   // Given diminishing returns function f(x) = x / (1+x),
+   // compute f(t/k).  The result is f(t/k) = (t/k) / (1+t/k),
+   // but we can multiply top and bottom by k to get
+   // t / (k + t).
+
+   // TODO:  Overflow analysis of this function
+   uint128_t dt_ms_prod = uint128_t(dt_ms) * DR_BONUS;
```
Comment (theoreticalbts) [inline contracts/pow/pow.cpp:174]: Again a TODO comment that needs an issue

File: contracts/pow/pow.cpp [hunk 6]
```diff
@@ -23,70 +29,185 @@ struct pow_signature_data
...
+}
+
+void update_difficulty( difficulty_metadata& diff_meta, timestamp_type current_block_time )
+{
+   uint64_t dt = uint64_t(current_block_time) - uint64_t(diff_meta.last_block_time);
+
+   // The order of the following computations assumes that the presented work
+   // is checked against the difficulty target before this function is called.
+
+   // TODO:  Overflow analysis of these computations
```
Comment (theoreticalbts) [inline contracts/pow/pow.cpp:195]: Another TODO comment that needs an issue number

File: contracts/pow/pow.cpp [hunk 7]
```diff
@@ -110,13 +231,6 @@ int main()
       system::exit_contract( 0 );
    }
 
-   if ( system::get_caller().caller_privilege != chain::privilege::kernel_mode )
-   {
-      system::print( "pow contract must be called from kernel" );
-      system::set_contract_return( false );
-      system::exit_contract( 0 );
-   }
-
```
Comment (theoreticalbts) [inline contracts/pow/pow.cpp:119]: Why is it safe to delete this check?

**Review by theoreticalbts — CHANGES_REQUESTED** [review]

**Comment by theoreticalbts** [comment]
I filed issue #20 for the overflow TODO's, just reference the issue in the TODO comments.

**Comment by sgerbino** [comment]
Usurped by #23.

**Pull request closed without merging**